File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix[0])
        for i in range(n // 2):  # 테두리 개수  
            for j in range(i, n-1-i):  # 현재 테두리에서 회전 가능한 범위
                tmp = matrix[n-1-j][i]
                matrix[n-1-j][i] = matrix[n-1-i][n-j-1]
                matrix[n-1-i][n-j-1] = matrix[j][n-1-i]
                matrix[j][n-1-i] = matrix[i][j]
                matrix[i][j] = tmp

File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();

        // step 1 - transpose matrix
        for (int i = 0; i < n; i++)
        {
            for (int j = i + 1; j < n; j++) // avoid same index between i and j
            {
                swap(matrix[i][j], matrix[j][i]);
            }
        }

        // step 2 - revserse each rows
        // 90 degree = transpose -> row reverse
        // 180 =       row reverse -> column reverse
        // 270 =       row reverse -> transpose
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n / 2; j++)
            {
                swap(matrix[i][j], matrix[i][n - 1 - j]);
            }
        }
    }
};